### PR TITLE
Registries V2

### PR DESCRIPTION
**Pull request by alexyaang**
This PR is currently just for my own record keeping. It's hard to look at what i've changed / deleted / commented out since the change is pretty drastic. 

TODO: (in no particular order)
- [ ] Make all generic registries commands work
  - [x] connectRegistry
  - [x] copyRemoteImageDigest
  - [x] copyRemoteFullTag
  - [ ] deleteRemoteImage
  - [ ] deployImageToAzure
  - [ ] deployImageToAca
  - [X] disconnectRegistryProvider
  - [ ] registryHelp
  - [ ] logInToDockerCli
  - [ ] logOutOfDockerCli
  - [X] pullImageFromRepository
  - [X] pullRepository
  - [ ] reconnectRegistry

---
- [ ] MakedockerHub commands work
  - [ ] openDockerHubInBrowser
 
---
- [ ] Make all Azure commands work
  - [ ] buildImageInAzure
  - [ ] createAzureRegistry
  - [ ] deleteAzureRegistry
  - [ ] deleteAzureRepository
  - [ ] openInAzurePortal
  - [ ] runAzureTask
  - [ ] runFileAsAzureTask
  - [ ] untagAzureImage
  - [ ] viewAzureProperties
  - [ ] viewAzureTaskLogs

---
- [ ] Release extensibility package

### CODE REVIEW

File: src/commands/registries/connectRegistry.ts
Comment (alexyaang) [inline src/commands/registries/connectRegistry.ts:1]: New changes

File: src/commands/registries/copyRemoteImageDigest.ts
Comment (alexyaang) [inline src/commands/registries/copyRemoteImageDigest.ts:1]: New changes

File: src/commands/registries/disconnectRegistry.ts
Comment (alexyaang) [inline src/commands/registries/disconnectRegistry.ts:1]: New changes

File: src/commands/registries/logInToDockerCli.ts
Comment (alexyaang) [inline src/commands/registries/logInToDockerCli.ts:1]: New changes

File: src/commands/registries/pullImages.ts
Comment (alexyaang) [inline src/commands/registries/pullImages.ts:1]: New changes

File: src/extensionVariables.ts
Comment (alexyaang) [inline src/extensionVariables.ts:1]: New changes

File: src/tree/registerTrees.ts
Comment (alexyaang) [inline src/tree/registerTrees.ts:1]: New changes

File: src/tree/registries/getInformationFromRegistryItem.ts
Comment (alexyaang) [inline src/tree/registries/getInformationFromRegistryItem.ts:1]: New changes

File: package.json
Comment (alexyaang) [inline package.json:1]: New changes

File: src/commands/registries/disconnectRegistry.ts [hunk 1]
```diff
@@ -3,17 +3,10 @@
...
-    let cachedProvider: ICachedRegistryProvider | undefined;
-    if (node instanceof InvalidTreeItem) {
-        cachedProvider = <ICachedRegistryProvider>node.data;
-    } else if (node) {
-        cachedProvider = node.cachedProvider;
-    }
-    await ext.registriesRoot.disconnectRegistry(context, cachedProvider);
+
+export async function disconnectRegistry(context: IActionContext, node?: UnifiedRegistryItem<unknown>): Promise<void> {
+    await node.provider.onDisconnect?.();
```
Comment (bwateratmsft) [inline src/commands/registries/disconnectRegistry.ts:11]: Shouldn't this be calling `disconnect` on URTDP?

Comment (alexyaang) [reply src/commands/registries/disconnectRegistry.ts:11]: yup your're right

File: src/commands/registerLocalCommand.ts [hunk 2]
```diff
@@ -1,31 +1,31 @@
...
-import { l10n } from 'vscode';
-import { DockerExtensionKind, getVSCodeRemoteInfo } from '../utils/getVSCodeRemoteInfo';
-import { registerCommand } from './registerCommands';
+// import { IActionContext } from '@microsoft/vscode-azext-utils';
+// import { l10n } from 'vscode';
+// import { DockerExtensionKind, getVSCodeRemoteInfo } from '../utils/getVSCodeRemoteInfo';
+// import { registerCommand } from './registerCommands';
 
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export function registerLocalCommand(commandId: string, callback: (context: IActionContext, ...args: any[]) => any, debounce?: number): void {
```
[the remark this answers is not in the document]
Comment (alexyaang) [reply src/commands/registerLocalCommand.ts:12]: good catch, i think i was just rushing to make the project compile and commented out everything that threw an error, i think at the time `registerCommand` had errors or something. 

i'll add it back

File: src/commands/registries/logInToDockerCli.ts [hunk 3]
```diff
@@ -42,10 +33,10 @@ export async function logInToDockerCli(context: IActionContext, node?: RegistryT
                     client => client.login({
                         username: username,
                         passwordStdIn: true,
-                        registry: creds.registryPath,
+                        registry: '' // TODO: change this
```
Comment (bwateratmsft) [inline src/commands/registries/logInToDockerCli.ts:36]: We can change this to be the baseUrl's host now right?

File: src/tree/RefreshManager.ts [hunk 4]
```diff
@@ -301,7 +301,7 @@ export class RefreshManager extends vscode.Disposable {
                         callback = () => ext.networksRoot.refresh(context);
                         break;
                     case 'registries':
-                        callback = () => ext.registriesRoot.refresh(context);
+                        ext.registriesRoot.refresh(); //TODO: change this
```
Comment (bwateratmsft) [inline src/tree/RefreshManager.ts:304]: Shouldn't this be:
```suggestion
                        callback = () => ext.registriesRoot.refresh();
```

File: src/tree/registerTrees.ts [hunk 5]
```diff
@@ -42,13 +44,17 @@ export function registerTrees(): void {
...
+    const urtdp = new UnifiedRegistryTreeDataProvider(ext.context.globalState);
+    const genericRegistryV2DataProvider = new GenericRegistryV2DataProvider(ext.context);
+    urtdp.registerProvider(new GitHubRegistryDataProvider(ext.context));
+    urtdp.registerProvider(new DockerHubRegistryDataProvider(ext.context));
+    urtdp.registerProvider(new AzureRegistryDataProvider(ext.context));
+    urtdp.registerProvider(genericRegistryV2DataProvider);
+    ext.registriesRoot = urtdp;
+    ext.registriesTreeView = vscode.window.createTreeView('dockerRegistries', { treeDataProvider: urtdp });
+    ext.registriesTree = urtdp;
+    // TODO: add gitlab registry provider when the new extension is ready
```
Comment (bwateratmsft) [inline src/tree/registerTrees.ts:56]: GitLab's provider will be added by the other extension, no code will be needed in this extension for that

File: src/tree/registries/Azure/AzureRegistryDataProvider.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                contextValue: 'azuresubscription',
+                iconPath: vscode.Uri.joinPath(this.extensionContext.extensionUri, 'resources', 'azureSubscription.svg'),
+            });
+        } else {
+            return super.getTreeItem(element);
+        }
+    }
+
+    protected override getAuthenticationProvider(item: AzureRegistryItem): ACROAuthProvider {
+        const registryString = item.registryUri.toString();
```
Comment (bwateratmsft) [inline src/tree/registries/Azure/AzureRegistryDataProvider.ts:117]: Isn't `registryUri` removed in favor of `baseUrl`?

File: src/tree/registries/UnifiedRegistryTreeDataProvider.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
+import { RegistryDataProvider } from '@microsoft/vscode-docker-registries';
+import * as vscode from 'vscode';
+
+export interface UnifiedRegistryItem<T> {
+    provider: RegistryDataProvider<T>;
+    wrappedItem: T;
+    parent: UnifiedRegistryItem<T> | undefined;
+}
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (bwateratmsft) [inline src/tree/registries/UnifiedRegistryTreeDataProvider.ts:10]: This linter suppression may be able to be removed; the rules are different in the Docker extension repo

File: src/tree/registries/getInformationFromRegistryItem.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+export function getFullImageNameFromRegistryItem(node: UnifiedRegistryItem<CommonTag>): string {
+    const imageName = getImageNameFromRegistryItem(node);
+    if (!isRegistry(node.parent.parent?.wrappedItem)) {
+        throw new Error('Unable to get full image name');
+    }
+
+    const registryLabel = (node.parent.parent.wrappedItem as CommonRegistry).label;
+    switch (node.wrappedItem.additionalContextValues?.[0] ?? '') {
+        case 'registryV2Tag': {
+            const authority = Uri.parse(registryLabel).authority; // this should parse since we have validation logic
```
Comment (bwateratmsft) [inline src/tree/registries/getInformationFromRegistryItem.ts:27]: Isn't `baseUrl` available on `node.parent.parent.wrappedItem` or maybe even `wrappedItem`?

File: src/commands/registries/logInToDockerCli.ts (hunk 3 above)
Comment (alexyaang) [reply src/commands/registries/logInToDockerCli.ts:36]: ahhh yup good catch

File: src/tree/registerTrees.ts (hunk 5 above)
Comment (alexyaang) [reply src/tree/registerTrees.ts:56]: ohh that makes sense

File: src/tree/registries/Azure/AzureRegistryDataProvider.ts (hunk 6 above)
Comment (alexyaang) [reply src/tree/registries/Azure/AzureRegistryDataProvider.ts:117]: yeah you're right. this is fixed in branch `yangalex/azureRegistries`, you should see the changes in that PR

Comment (alexyaang) [reply src/tree/registries/Azure/AzureRegistryDataProvider.ts:117]: let em add the changes here too

File: src/tree/registries/getInformationFromRegistryItem.ts (hunk 8 above)
Comment (alexyaang) [reply src/tree/registries/getInformationFromRegistryItem.ts:27]: good catch 😃 , forgot to change this after adding the `baseUrl` field

**Review by bwateratmsft — APPROVED** [review]

**Pull request merged**